fix insertion before a node with multiple inputs + support additional broadcasting

jmduarte opened the pull request:
This PR has a few related fixes
- fix insertion before a node with multiple inputs
- correctly insert broadcast stream before a node with multiple inputs (e.g. Add or Concatenate), e.g. see figure below. The current code would behave incorrectly and replace the wrong input (always the first input) for the following layer... this would result in an additional output incorrectly being inferred
- don't insert broadcast before concatenate (figure below incorrectly does this)

<img width="782" alt="Screen Shot 2022-05-16 at 6 59 26 PM" src="https://user-images.githubusercontent.com/4932543/168712933-502f31c8-7a2c-4b69-883c-74e10f4d534f.png">

To do
- [x] Add a pytest

@nickshey

thesps commented:
Pushed to local branch pr/551 to run Gitlab CI, looks like the graph tests still pass. An addition to `test_graph.py` that touches this new `input_idx` argument would be good.

jmduarte commented:
@thesps I added the test you requested. It was failing though because I realized the only stream broadcasting we currently supported as of the form: `(1, 1, C) -> (H, W, C)`, and I was testing a different form, namely `(H, W, 1) -> (H, W, C)`.

So I updated our HLS code to support both of these (easy) cases. I also added tests for both cases, and added a `RuntimeError` if a user tries to use a model with broadcasting that is not one of these two cases.


In principle, it should be straightforward(ish) to also support `(1, W, C) -> (H, W, C)` and `(H, 1, C) -> (H, W, C)`, but this can be a future PR

jmduarte commented:
@thesps I think this is ready to be merged unless you have any other comments!
